### DISCUSSION TITLE
How to update scene graph at runtime?

### DISCUSSION THREAD

**Discussion by yangcshen** [vsg-dev/VulkanSceneGraph#805] [General]
Hi

I am still getting up to speed with vsg api. I have a general question about how to update scene graph at runtime. 
Here is a scenario.
Let's say I am using ImGui for UI. I created a button which will create a sphere using vsg::builder and add it to the scene. This needs to happen at runtime instead of initialization of the application, meaning the code gets executed at the current frame and the updates will be reflected on the next frame.

1. I have seen vsgdynamicload and vsgdynamicviews examples which use vsg::Operation and vsg::OperationThreads to update scene graph at runtime. My understanding is that if there is anything to update related to the scene graph, this is the pattern to use. Is this correct? What if I want to put everything on the main thread? vsg_scene->addNode() -> viewer->compile() ->  updateViewer()?
2. I see the examples always implement an operation struct and a merge struct? Can I combine them into one? What is the reason to have separate structs? 

Thanks,
Cam

**Comment by robertosfield** [comment]
Hi Cam,

The use of vsg::Operation is only really needed for multi-threaded usage, for the case of responding to UI this can normally be done single threaded and as update of the update phase of the scene graph.  The vsgintersection example illustrates this type of usage.

https://github.com/vsg-dev/vsgExamples/blob/master/examples/utils/vsgintersection/vsgintersection.cpp

**Reply by yangcshen** [reply]
Hi Robert,

Thanks for your reply. I am little confused about dealing with UI events using ImGui. If I implement a handler inherited from vsg::Visitor,     `void apply(vsg::ButtonPressEvent& buttonPressEvent) override` will invoke first and then the event will be passed to ImGui layer. But my event flow should be the opposite. ImGui layer receives user event first and then my handler does `apply`. Please advise. Thanks.

**Reply by robertosfield** [reply]
If you are using ImGui then the vsg::Visitor isn't appropriate, you'll need to respond to the even within your ImGui code.

**Reply by yangcshen** [reply]
I tried to put my event logics in ImGui code but when the main loop hits `viewer->recordAndSubmit();`, it will pop out the following error message.

```
File: C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\vector
Line: 1959

Expression: vector subscript out of range

For information on how your program can cause an assertion
failure, see the Visual C++ documentation on asserts.
```
This error stems from https://github.com/vsg-dev/VulkanSceneGraph/blob/1fcd53589ea34b3595d32125af5dc4b58fb28b04/src/vsg/app/CommandGraph.cpp#L115

**Reply by robertosfield** [reply]
There isn't any way for us to know what might be wrong as we haven't seen any of your code, all we know is you've done something with ImGui that is someway related to the scene graph.

Could you modify an example like vsgimgui to reproduce your usage case so we can see how you are approaching things and ideally be able to run it to see the error for ourselves.

**Reply by yangcshen** [reply]
Hi Robert, 

I made a stupid mistake, where I created a different vsg::Builder instance and used that instance to addNode. All is good now. 
Thanks for your prompt reply.

Cam
